### PR TITLE
Feature/flatfile tests

### PR DESCRIPTION
**Pull request by x3medima17**
### Description of the Change
This PR covers entire FlatFile class with tests. 

### Benefits
Maximum coverage.

### Possible Drawbacks 
None.

### CODE REVIEW

File: irohad/ametsuchi/impl/flat_file/flat_file.cpp [hunk 1]
```diff
@@ -94,23 +95,21 @@ namespace {
 std::unique_ptr<FlatFile> FlatFile::create(const std::string &path) {
```
Comment (Warchant) [inline irohad/ametsuchi/impl/flat_file/flat_file.cpp]: Why uptr?
Should it be optional?

File: test/module/irohad/ametsuchi/flat_file_test.cpp [hunk 2]
```diff
@@ -86,8 +90,9 @@ TEST_F(BlStore_Test, BlockStoreWhenRemoveBlock) {
 }
 
 TEST_F(BlStore_Test, BlockStoreWhenAbsentFolder) {
-  log_->info("----------| Check that folder absent => create => "
-                 "make storage => remove storage |----------");
+  log_->info(
+      "----------| Check that folder absent => create => "
```
Comment (Warchant) [inline test/module/irohad/ametsuchi/flat_file_test.cpp]: Check that folder **is** absent

File: test/module/irohad/ametsuchi/flat_file_test.cpp [hunk 3]
```diff
@@ -60,8 +63,9 @@ TEST_F(BlStore_Test, BlockStoreWhenRemoveBlock) {
   log_->info("----------| Simulate removal of the block |----------");
   // Remove file in the middle of the block store
   {
-    log_->info("----------| create blockstore and insert 3 elements "
-                   "|----------");
+    log_->info(
+        "----------| create blockstore and insert 3 elements "
```
Comment (Warchant) [inline test/module/irohad/ametsuchi/flat_file_test.cpp]: is it really necessary to write these decorators?

e.g. `----------| something |----------`

File: test/module/irohad/ametsuchi/flat_file_test.cpp [hunk 4]
```diff
@@ -122,3 +127,62 @@ TEST_F(BlStore_Test, BlockStoreInitializationFromNonemptyFolder){
...
+  std::vector<uint8_t> block(100000, 5);
+  auto bl_store = FlatFile::create("");
+  ASSERT_EQ(bl_store, nullptr);
+}
+
+TEST_F(BlStore_Test, WriteDeniedFolder) {
+  std::vector<uint8_t> block(100000, 5);
+  auto bl_store = FlatFile::create(block_store_path);
+  auto id = 1u;
+  chmod(block_store_path.data(), S_IRUSR | S_IXUSR);
```
Comment (Warchant) [inline test/module/irohad/ametsuchi/flat_file_test.cpp]: It should be possible to do without posix calls, but with use of `boost`

File: test/module/irohad/ametsuchi/flat_file_test.cpp [hunk 5]
```diff
@@ -122,3 +127,62 @@ TEST_F(BlStore_Test, BlockStoreInitializationFromNonemptyFolder){
...
+  auto res = bl_store->add(id, block);
+  ASSERT_FALSE(res);
+  chmod(block_store_path.data(), S_IRWXU | S_IRWXG | S_IROTH | S_IXOTH);
```
Comment (Warchant) [inline test/module/irohad/ametsuchi/flat_file_test.cpp]: And here

File: test/module/irohad/ametsuchi/flat_file_test.cpp [hunk 6]
```diff
@@ -122,3 +127,62 @@ TEST_F(BlStore_Test, BlockStoreInitializationFromNonemptyFolder){
...
+}
+
+TEST_F(BlStore_Test, GetDeniedBlock) {
+  std::vector<uint8_t> block(100000, 5);
+  auto bl_store = FlatFile::create(block_store_path);
+  auto id = 1u;
+  bl_store->add(id, block);
+
+  auto filename = boost::filesystem::path{block_store_path} / id_to_name(id);
+  chmod(filename.string().data(), 0);
```
Comment (Warchant) [inline test/module/irohad/ametsuchi/flat_file_test.cpp]: Posix -> Boost

File: irohad/ametsuchi/impl/flat_file/flat_file.cpp (hunk 1 above)
Comment (x3medima17) [reply irohad/ametsuchi/impl/flat_file/flat_file.cpp]: It was initially unique_ptr

File: test/module/irohad/ametsuchi/flat_file_test.cpp (hunk 3 above)
Comment (x3medima17) [reply test/module/irohad/ametsuchi/flat_file_test.cpp]: It was initially written like this

File: test/module/irohad/ametsuchi/flat_file_test.cpp (hunk 2 above)
Comment (x3medima17) [reply test/module/irohad/ametsuchi/flat_file_test.cpp]: fixed

File: test/module/irohad/ametsuchi/flat_file_test.cpp (hunk 6 above)
Comment (x3medima17) [reply test/module/irohad/ametsuchi/flat_file_test.cpp]: Fixed

File: test/module/irohad/ametsuchi/flat_file_test.cpp (hunk 4 above)
Comment (x3medima17) [reply test/module/irohad/ametsuchi/flat_file_test.cpp]: Fixed

File: test/module/irohad/ametsuchi/flat_file_test.cpp (hunk 5 above)
Comment (x3medima17) [reply test/module/irohad/ametsuchi/flat_file_test.cpp]: Fixed

File: test/module/irohad/ametsuchi/flat_file_test.cpp [hunk 7]
```diff
@@ -16,6 +16,8 @@
  */
 
 #include "ametsuchi/impl/flat_file/flat_file.hpp"
+#include "ametsuchi/impl/flat_file/flat_file.cpp"
```
Comment (l4l) [inline test/module/irohad/ametsuchi/flat_file_test.cpp]: Something went wrong here

File: test/module/irohad/ametsuchi/flat_file_test.cpp [hunk 8]
```diff
@@ -122,3 +126,71 @@ TEST_F(BlStore_Test, BlockStoreInitializationFromNonemptyFolder){
   // check that last ids of both block storages are the same
   ASSERT_EQ(bl_store1->last_id(), bl_store2->last_id());
 }
+
+TEST_F(BlStore_Test, EmptyDumpDir) {
+  auto res = check_consistency("");
+  ASSERT_EQ(res, nonstd::nullopt);
```
Comment (l4l) [inline test/module/irohad/ametsuchi/flat_file_test.cpp]: just `ASSERT_FALSE(res)`
also look below

File: test/module/irohad/ametsuchi/flat_file_test.cpp [hunk 9]
```diff
@@ -122,3 +126,71 @@ TEST_F(BlStore_Test, BlockStoreInitializationFromNonemptyFolder){
...
+}
+
+TEST_F(BlStore_Test, GetNonExistingFile) {
+  auto bl_store = FlatFile::create(block_store_path);
+  Identifier id = 98759385;  // random number that will not exist
```
Comment (l4l) [inline test/module/irohad/ametsuchi/flat_file_test.cpp]: s/will/does/

File: test/module/irohad/ametsuchi/flat_file_test.cpp [hunk 10]
```diff
@@ -122,3 +126,71 @@ TEST_F(BlStore_Test, BlockStoreInitializationFromNonemptyFolder){
...
+
+  auto filename = boost::filesystem::path{block_store_path} / id_to_name(id);
+
+  using perms = boost::filesystem::perms;
+  auto curr = boost::filesystem::status(boost::filesystem::path(filename));
+
+  boost::filesystem::permissions(filename.string().data(), perms::no_perms);
+  auto res = bl_store->get(id);
+  ASSERT_EQ(res, nonstd::nullopt);
+  boost::filesystem::permissions(filename.string().data(), curr.permissions());
```
Comment (l4l) [inline test/module/irohad/ametsuchi/flat_file_test.cpp]: What's the point of that call? The file is cleaned-up after the test

File: test/module/irohad/ametsuchi/flat_file_test.cpp [hunk 11]
```diff
@@ -122,3 +126,71 @@ TEST_F(BlStore_Test, BlockStoreInitializationFromNonemptyFolder){
...
+}
+
+TEST_F(BlStore_Test, AddExistingId) {
+  std::vector<uint8_t> block(100000, 5);
+  auto bl_store = FlatFile::create(block_store_path);
+  auto id = 1u;
+  const auto file_name =
+      boost::filesystem::path{block_store_path} / id_to_name(id);
+  std::ofstream fout(file_name.string());
```
Comment (l4l) [inline test/module/irohad/ametsuchi/flat_file_test.cpp]: Are this implying file truncation? If so pls specify such flag explicitly

File: test/module/irohad/ametsuchi/flat_file_test.cpp [hunk 12]
```diff
@@ -122,3 +126,71 @@ TEST_F(BlStore_Test, BlockStoreInitializationFromNonemptyFolder){
...
+  ASSERT_EQ(res, nonstd::nullopt);
+}
+
+TEST_F(BlStore_Test, GetDirectory) {
+  auto bl_store = FlatFile::create(block_store_path);
+  ASSERT_EQ(bl_store->directory(), block_store_path);
+}
+
+TEST_F(BlStore_Test, GetDeniedBlock) {
+  std::vector<uint8_t> block(100000, 5);
```
Comment (l4l) [inline test/module/irohad/ametsuchi/flat_file_test.cpp]: Used pretty often, so you may extract it to the fixture (and rename to blob or something like that)

File: test/module/irohad/ametsuchi/flat_file_test.cpp [hunk 13]
```diff
@@ -122,3 +126,71 @@ TEST_F(BlStore_Test, BlockStoreInitializationFromNonemptyFolder){
...
+
+  using perms = boost::filesystem::perms;
+  auto path = boost::filesystem::path(block_store_path);
+  auto curr = boost::filesystem::status(path);
+
+  boost::filesystem::permissions(path, perms::owner_read | perms::owner_exe);
+  auto res = bl_store->add(id, block);
+  ASSERT_FALSE(res);
+  boost::filesystem::permissions(path, curr.permissions());
+}
```
Comment (l4l) [inline test/module/irohad/ametsuchi/flat_file_test.cpp]: missing newline :(

File: test/module/irohad/ametsuchi/flat_file_test.cpp (hunk 7 above)
Comment (x3medima17) [reply test/module/irohad/ametsuchi/flat_file_test.cpp]: I need check_consistency function, it is declared in cpp file

Comment (l4l) [reply test/module/irohad/ametsuchi/flat_file_test.cpp]: Thus maybe move this function declaration to the hpp?

File: test/module/irohad/ametsuchi/flat_file_test.cpp (hunk 11 above)
Comment (x3medima17) [reply test/module/irohad/ametsuchi/flat_file_test.cpp]: Nope, this just creates an empty file, so add() will fail

File: test/module/irohad/ametsuchi/flat_file_test.cpp (hunk 7 above)
Comment (x3medima17) [reply test/module/irohad/ametsuchi/flat_file_test.cpp]: fixed

File: test/module/irohad/ametsuchi/flat_file_test.cpp (hunk 9 above)
Comment (x3medima17) [reply test/module/irohad/ametsuchi/flat_file_test.cpp]: Fixed

File: test/module/irohad/ametsuchi/flat_file_test.cpp (hunk 12 above)
Comment (x3medima17) [reply test/module/irohad/ametsuchi/flat_file_test.cpp]: Fixed

File: test/module/irohad/ametsuchi/flat_file_test.cpp (hunk 13 above)
Comment (x3medima17) [reply test/module/irohad/ametsuchi/flat_file_test.cpp]: Fixed

File: test/module/irohad/ametsuchi/flat_file_test.cpp (hunk 10 above)
Comment (l4l) [reply test/module/irohad/ametsuchi/flat_file_test.cpp]: screenshot_20180114_000649
not really

Comment (x3medima17) [reply test/module/irohad/ametsuchi/flat_file_test.cpp]: Fixed

**Comment by neewy** [comment]
>Maximum coverage.
>Merging #855 into develop will increase coverage by 0.03%.

😭

File: irohad/ametsuchi/impl/flat_file/flat_file.cpp [hunk 14]
```diff
@@ -21,96 +21,80 @@
...
 #include <boost/range/algorithm/find_if.hpp>
 #include <iomanip>
+#include <iostream>
 #include <sstream>
 #include "common/files.hpp"
 
-using namespace iroha::ametsuchi;
+namespace iroha {
+  namespace ametsuchi {
+    const uint32_t DIGIT_CAPACITY = 16;
```
Comment (muratovv) [inline irohad/ametsuchi/impl/flat_file/flat_file.cpp]: There is an idea for improving coverage. You should move `DIGIT_CAPACITY` into the class as a static constant field. Pls, check the coverage before/after and make a report about it.

File: irohad/ametsuchi/impl/flat_file/flat_file.cpp [hunk 15]
```diff
@@ -21,96 +21,80 @@
 #include <boost/range/adaptor/indexed.hpp>
 #include <boost/range/algorithm/find_if.hpp>
 #include <iomanip>
+#include <iostream>
 #include <sstream>
 #include "common/files.hpp"
 
-using namespace iroha::ametsuchi;
+namespace iroha {
+  namespace ametsuchi {
```
Comment (muratovv) [inline irohad/ametsuchi/impl/flat_file/flat_file.cpp]: What is the reason to use namespaces inside `cpp` files?

File: irohad/ametsuchi/impl/flat_file/flat_file.cpp [hunk 16]
```diff
@@ -21,96 +21,80 @@
...
-    if (!boost::filesystem::is_directory(path)) {
-      log_->error("Cannot create storage dir: {}", path);
-    }
-  }
-  auto res = check_consistency(path);
-  if (not res) {
-    log_->error("Checking consistency for {} - failed", path);
+  boost::system::error_code err;
+  if (not boost::filesystem::is_directory(path, err)
+      and not boost::filesystem::create_directory(path, err)) {
```
Comment (muratovv) [inline irohad/ametsuchi/impl/flat_file/flat_file.cpp]: Nice fix

File: irohad/ametsuchi/impl/flat_file/flat_file.cpp (hunk 1 above)
Comment (muratovv) [reply irohad/ametsuchi/impl/flat_file/flat_file.cpp]: I think, also, we need to rework return value from `create()` with optional type. 

The reason is this file not satisfy our policy about pointers.

File: irohad/ametsuchi/impl/flat_file/flat_file.cpp [hunk 17]
```diff
@@ -21,96 +21,80 @@
...
     return nullptr;
   }
-  return std::unique_ptr<FlatFile>(new FlatFile(*res, path));
+
+  auto res = check_consistency(path);
+  return std::make_unique<FlatFile>(*res, path, private_tag{});
 }
 
-void FlatFile::add(Identifier id, const std::vector<uint8_t> &block) {
+bool FlatFile::add(Identifier id, const std::vector<uint8_t> &block) {
```
Comment (muratovv) [inline irohad/ametsuchi/impl/flat_file/flat_file.cpp]: In nearest future, we want to rework interfaces of storage with `Result` type. So, I request to write todo with reworking method with respect to this behavior.

**Comment by x3medima17** [comment]
>Maximum coverage.
>Merging #855 into develop will increase coverage by 0.03%.

>😭

image

File: test/module/irohad/ametsuchi/flat_file_test.cpp (hunk 3 above)
Comment (muratovv) [reply test/module/irohad/ametsuchi/flat_file_test.cpp]: If you are reworking logging string better to remove them and write `given`, `when`, `then` notation.

File: test/module/irohad/ametsuchi/flat_file_test.cpp [hunk 18]
```diff
@@ -122,3 +127,64 @@ TEST_F(BlStore_Test, BlockStoreInitializationFromNonemptyFolder){
   // check that last ids of both block storages are the same
   ASSERT_EQ(bl_store1->last_id(), bl_store2->last_id());
 }
+
+TEST_F(BlStore_Test, EmptyDumpDir) {
```
Comment (muratovv) [inline test/module/irohad/ametsuchi/flat_file_test.cpp]: Same with `GWZ` for the new tests

File: test/module/irohad/ametsuchi/flat_file_test.cpp [hunk 19]
```diff
@@ -122,3 +127,64 @@ TEST_F(BlStore_Test, BlockStoreInitializationFromNonemptyFolder){
...
+
+  using perms = boost::filesystem::perms;
+
+  boost::filesystem::permissions(filename.string().data(), perms::no_perms);
+  auto res = bl_store->get(id);
+  ASSERT_EQ(res, nonstd::nullopt);
+}
+
+TEST_F(BlStore_Test, AddExistingId) {
+  std::vector<uint8_t> block(100000, 5);
```
Comment (muratovv) [inline test/module/irohad/ametsuchi/flat_file_test.cpp]: Numbers look not so obvious to own self. Maybe better to refactor it with variables?

File: test/module/irohad/ametsuchi/flat_file_test.cpp (hunk 3 above)
Comment (x3medima17) [reply test/module/irohad/ametsuchi/flat_file_test.cpp]: I just ran CLang Format :D 
I did not change anything

**Pull request merged**